mlrun/runtimes/base.py

```diff
@@ -1284,6 +1284,86 @@ def monitor_runs(self, db: DBInterface, db_session: Session):
                     exc=str(exc),
                     traceback=traceback.format_exc(),
                 )
+        for project, runs in project_run_uid_map.items():
+            if runs:
+                for run_uid in runs:
+                    self._ensure_run_not_stuck_on_non_terminal_state(
+                        db, db_session, project, run_uid, runtime_resources
+                    )
+
+    def _ensure_run_not_stuck_on_non_terminal_state(
+        self, db: DBInterface, db_session: Session, project, run_uid, runtime_resources
+    ):
+        """
+        Assuring that a run does not become trapped in a non-terminal state as a result of not finding
+        corresponding k8s resource.
+        This can occur when a node is evicted or preempted, causing the resources to be removed from the resource
+        listing when the final state recorded in the database is non-terminal.
+        This will have a significant impact on scheduled jobs, since they will not be created until the
+        previous run reaches a terminal state
+        """
+        try:
+            run = db.read_run(db_session, run_uid, project)
+        except mlrun.errors.MLRunNotFoundError:
+            logger.warning(
+                "Run not found, got deleted by another process",
+                project=project,
+                uid=run_uid,
+            )
+            return
+        now = now_date()
+        db_run_state = run.get("status", {}).get("state")
+        if not db_run_state:
+            # can't do any verification due to no run_state updated
+            return
+        if db_run_state in RunStates.k8s_non_terminal_states():
+            for runtime_resource in runtime_resources:
```

Comment:
Talked offline, performance - the list can be long, let's build a map to allow ourselves to search efficiently